semaphore.cpp

```diff
@@ -1,23 +1,30 @@
 #include "semaphore.h"
-
+#include <cassert>
 
 Semaphore::Semaphore(int initialValue)
     :mSemaphoreValue(initialValue)
 {
-
+    assert(initialValue >= 0);
 }
 
-void Semaphore::Increment() //Signal - Unlock
+void Semaphore::Signal() //Signal - Unlock
 {
+    std::lock_guard<std::mutex> lck(mMutexObject);
     mSemaphoreValue++;
-    mMutexObject.unlock();
+    if (mSemaphoreValue == 1)
+    {
+        mConditionVariable.notify_one();
+    }
 }
 
-void Semaphore::Decrement() //Wait - Lock
+void Semaphore::Wait() //Wait - Lock
 {
-    mSemaphoreValue--;
-    if (mSemaphoreValue < 0)
+    std::unique_lock<std::mutex> lck(mMutexObject);
+    while (mSemaphoreValue < 0)
```

Comment:
What if it's equal to 0?

Reply:
Will change condition to while (mSemaphoreValue < 0) :ok: